artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/AMQPConnectionContext.java

```diff
@@ -416,12 +411,15 @@ public void onRemoteOpen(Connection connection) throws Exception {
        * */
       if (connection.getRemoteProperties() == null || !connection.getRemoteProperties().containsKey(CONNECTION_OPEN_FAILED)) {
          long nextKeepAliveTime = handler.tick(true);
-         synchronized (schedulingLock) {
-            if (nextKeepAliveTime != 0 && scheduledPool != null && !isSchedulingCancelled) {
-               scheduledFuture = scheduledPool.schedule(new ScheduleRunnable(), (nextKeepAliveTime - TimeUnit.NANOSECONDS.toMillis(System.nanoTime())), TimeUnit.MILLISECONDS);
+
+         scheduledFutureRef.getAndUpdate(value -> {
+            if (value != null && nextKeepAliveTime != 0 && scheduledPool != null) {
```

Comment:
nextKeepAliveTime != 0 && scheduledPool != null can be pulled outside the atomic function call.

Reply:
I completely agree I will pull the "nextKeepAliveTime != 0 && scheduledPool != null" out outside.

Reply:
still waiting on this change.

Reply:
```
      if (scheduledPool != null && (connection.getRemoteProperties() == null || !connection.getRemoteProperties().containsKey(CONNECTION_OPEN_FAILED))) {
         long nextKeepAliveTime = handler.tick(true);
         if (nextKeepAliveTime != 0) {
            scheduledFutureRef.getAndUpdate(value -> {
               if (value != null) {
                  return scheduledPool.schedule(new ScheduleRunnable(), (nextKeepAliveTime - TimeUnit.NANOSECONDS.toMillis(System.nanoTime())), TimeUnit.MILLISECONDS);
               }

               return value;
            });
         }
      }
```